docs: document contract architecture and failure scenarios (#430)
FilOzone/filecoin-services: SPEC.md

### Service Termination

When terminating a dataset's service, FWSS terminates the PDP rail (which it validates) and performs best-effort termination of CDN rails, ignoring any errors. This ensures service termination succeeds regardless of CDN rail state—whether rails are active, already terminated, or fully settled and finalized.

## Contract Architecture

FWSS is composed of multiple independently-deployed contracts connected by on-chain references. Understanding the coupling between these contracts is essential for reasoning about upgradeability, failure recovery, and migration.

> ### System Diagram

Reviewer: I turned this into a mermaid diagram.  Maybe a little cleaner?  I can either way?

```mermaid
flowchart TB
  PDPV["PDPVerifier (UUPS proxy)"]
  FWSS["FilecoinWarmStorageService (UUPS proxy)"]
  FPV["FilecoinPayV1 (immutable)"]
  SPR["ServiceProviderRegistry (UUPS proxy)"]
  SKR["SessionKeyRegistry (immutable)"]
  SVL["SignatureVerificationLib (library)"]
  SV["StateView (immutable)"]
  USDFC["USDFC Token (immutable ERC20)"]

  PDPV -->|PDPListener callbacks| FWSS
  FWSS -->|createRail, modifyRailPayment| FPV
  FPV -->|IValidator: validatePayment, railTerminated| FWSS
  FWSS -->|getProviderIdByAddress, getProviderPayee| SPR
  FWSS -->|uses| SVL
  SVL -->|authorizationExpiry| SKR
  SV -->|read-only queries| FWSS
  FWSS -.->|references| USDFC
  FPV -.->|references| USDFC
```

**[UUPS]** (Universal Upgradeable Proxy Standard, [ERC-1822](https://eips.ethereum.org/EIPS/eip-1822)) is a proxy pattern where the upgrade logic lives in the implementation contract rather than the proxy. The proxy delegates all calls to the current implementation, and only the implementation can authorize a switch to a new one. See the [OpenZeppelin UUPS guide](https://docs.openzeppelin.com/contracts/5.x/api/proxy#UUPSUpgradeable) for details.

### Upgrade Mechanism

FWSS, PDPVerifier, and ServiceProviderRegistry each implement the same **two-step announce-then-upgrade** pattern. The owner first calls `announcePlannedUpgrade(PlannedUpgrade)` with the address of the new implementation and a future block number. The upgrade cannot execute until that block number is reached. When `upgradeToAndCall` is later invoked, the internal `_authorizeUpgrade` guard verifies the implementation matches the announcement and the delay has elapsed:

```solidity
function announcePlannedUpgrade(PlannedUpgrade calldata plannedUpgrade) external onlyOwner {
    require(plannedUpgrade.nextImplementation.code.length > 3000);

>     require(plannedUpgrade.afterEpoch > block.number);

Reviewer: ```suggestion
    require(plannedUpgrade.afterEpoch > block.number); // plannedUpgrade.afterEpoch  must be at least one block later than `block.number`.  The delay is intended to gives users time to review the announced implementation before it takes effect.
```

    nextUpgrade = plannedUpgrade;
    emit UpgradeAnnounced(plannedUpgrade);
}

function _authorizeUpgrade(address newImplementation) internal override onlyOwner {
    require(newImplementation == nextUpgrade.nextImplementation);
    require(block.number >= nextUpgrade.afterEpoch);
    delete nextUpgrade;
}
```

> See [`FilecoinWarmStorageService.sol:429-441`](https://github.com/FilOzone/filecoin-services/blob/main/service_contracts/src/FilecoinWarmStorageService.sol#L429-L441). The minimum code-length check (`code.length > 3000`) prevents accidental upgrades to empty or trivially small implementations. The `afterEpoch` delay gives users time to review the announced implementation before it takes effect.

Reviewer: ```suggestion
_(This is from [`FilecoinWarmStorageService.sol:429-441`](https://github.com/FilOzone/filecoin-services/blob/main/service_contracts/src/FilecoinWarmStorageService.sol#L429-L441).)_

A successful `_authorizeUpgrade` call will enable the old implementation to set the `newImplementation` address on the proxy, effectively completing the upgrade.
```

I inlined some of your original text as code comments.
I also wanted to close the loop on explianing how we go from the proxy pointing to the old address to the new address.

**TODO:** Evaluate whether to add a pause mechanism to ServiceProviderRegistry to prevent new registrations during an incident.

### SessionKeyRegistry

SessionKeyRegistry is **not upgradeable**. It stores session key authorization mappings used for signature delegation.

**Impact assessment.** The impact of a SessionKeyRegistry flaw is relatively low compared to FilecoinPayV1 because it holds only authorization state, not funds. Replacing it requires a new FWSS implementation pointing to the new registry (plus the two-step upgrade). Users must re-register their session keys in the new contract. Existing datasets and rails are unaffected — session keys are only checked during dataset creation and piece operations, not during settlement or proving.

**TODO:** Evaluate a dual-registry transition period where FWSS checks both the old and new registry during migration, allowing users to migrate keys without a hard cutover.

> 

Reviewer: ```suggestion

### StateView (FilecoinWarmStorageServiceStateView)

* StateView is an **immutable** read-only contract that queries FWSS state through the FWSS proxy. 
* FWSS does not hold a reference to the StateView. 
* StateView is reliant on no breaking changes to FWSS.
* **If this piece fails:** Redeploy StateView with fixed logic. No FWSS upgrade is required because callers (e.g., Synapse, UIs) point at StateView directly. Consumers like Synapse need to be updated to point to a new StateView contract.  

```

Reviewer 2: `viewContractAddress` is mutable in FWSS, that should be mentioned in this section, we deploy a new StateView, update that address and it _should_ work. But in practice, Synapse is now hard-wiring the StateView address and will just bump that when we deploy a new version and relese a new Synapse version pointing to the new one, this is ~harmless unless we find some critical bug (which is fairly unlikely becuause it's such a simple contract)

**Proxy vs. non-proxy references.** References to proxy contracts (PDPVerifier, ServiceProviderRegistry) are **resilient**: those contracts upgrade their own logic independently without requiring any change to FWSS. The FWSS immutable address points at the proxy, which remains stable across upgrades.

References to non-proxy contracts (FilecoinPayV1, SessionKeyRegistry) are **rigid**: replacing them requires a new FWSS implementation carrying the new address, plus the two-step upgrade. State held in the old contract (rail balances, lockups, session key authorizations) is **orphaned** — it cannot be migrated automatically and must be handled explicitly.

### Data Flows

The following sequences describe the primary cross-contract call paths.

```mermaid
sequenceDiagram
  participant Caller as Any Caller
  participant Client
  participant SP as Storage Provider
  participant PDPVerifier
  participant FWSS
  participant ServiceProviderRegistry
  participant FilecoinPayV1

  rect rgba(248, 248, 248, 0.2)
    Note over Client: 1. Dataset Creation
    Client->>PDPVerifier: create dataset
    PDPVerifier->>FWSS: dataSetCreated callback

>     FWSS->>ServiceProviderRegistry: getProviderPayee(providerId)

Reviewer 2: Not `Client`, but `SP` calls this like in AddPieces below. But also, we should put `Client->>SP` in here and in AddPieces. Also there's an EIP-712 signature check in here since `extraData` flows from Client to SP to PDPVerifier to FWSS where we validate it, same as AddPieces.

Also we are missing ScheduleDeletions in this, and that's a little more complicated because it's not an actual deletion, it's a schedule deletion and the deletion happens on nextProvingPeriod`.

- **Backwards incompatibility of implementation contracts**: Removing or changing the signature of a public/external function in a new implementation breaks callers that depend on the old ABI — most critically FilecoinPayV1, which is immutable and calls FWSS through the `IValidator` interface. The codebase checks in ABI JSON files under [`service_contracts/abi/`](https://github.com/FilOzone/filecoin-services/tree/main/service_contracts/abi) and CI ([`check-abi`](https://github.com/FilOzone/filecoin-services/blob/main/.github/workflows/check.yml)) regenerates them on every run, failing if any ABI has changed. This ensures that any function addition, removal, or signature change is surfaced as a diff during code review. As with storage layout, this is a **detection** mechanism — it flags changes but does not structurally prevent a breaking removal.

> **TODO:** Define concrete procedures for coordinated multi-contract upgrades. Document a pre-upgrade testing checklist. Specify rollback procedures and emergency response timelines. Evaluate whether to adopt structural protections such as OpenZeppelin storage gaps or ERC-7201 namespaced storage.

Reviewer 3: There are two other standards for storage domains. One is [ERC-8042](https://eips.ethereum.org/EIPS/eip-8042) and another is [Ithaca](https://github.com/ithacaxyz/account/blob/ab0a493f04676abc71104623bcfcdede050da2ec/src/IthacaAccount.sol#L102).

> **TODO:** Define concrete procedures for coordinated multi-contract upgrades. Document a pre-upgrade testing checklist. Specify rollback procedures and emergency response timelines. Evaluate whether to adopt structural protections such as OpenZeppelin storage gaps or ERC-7201 namespaced storage.

Author: Addressed in: https://github.com/FilOzone/filecoin-services/pull/430/changes/8e8b89f4eaf441ab5fd1467b86e556d1d073ad46
